test/functional/testlib.bash

```diff
@@ -775,6 +787,118 @@ destroy_test_environment() {
 
 }
 
+# creates a web server to host fake swupd content with or without certifiates
+start_web_server() {
+
+	local port
+	local server_args
+	local server_pid
+	local status
+
+	cb_usage() { 
+		echo $(cat <<-EOF
+		Usage:
+		    start_web_server [-k] <server priv key> [-p] <server pub key> [-s]
+
+		Options:
+		    -k    Path to server private key which must correspond to the provided server public key
+		    -p    Path to server public key which enables SSL authentication
+		    -s    Use a slow update server
+
+		Notes:
+		    - When the server is using SSL authentication, a pair of corresponding public and private keys must be passed
+		      as arguments.
+
+		Example of usage:
+
+		    The following command will create a web server that uses SSL authentication and inserts a delay for content
+		    updates.
+
+		    start_web_server -k /private_key.pem -p /public_key.pem -s
+
+		EOF
+		)
+	}
+
+	set -f  # turn off globbing
+	while getopts :k:p:s opt; do
+		case "$opt" in
+			k)	server_args="$server_args --server_key $OPTARG" ;;
+			p)	server_args="$server_args --server_cert $OPTARG" ;;
+			s)	server_args="$server_args --slow_server" ;;
+			*)	cb_usage
+				return ;;
+		esac
+	done
+	set +f  # turn globbing back on
+	
+	# get available port
+	for i in $(seq 8081 8181); do
+
+		# start server
+		python $FUNC_DIR/server.py $server_args --port $i &
+		server_pid=$!
+
+		sleep .5
```

Comment:
When running a lot of tests in parallel, a failure can occur here. If the server.py script uses an invalid port and does not fail before 0.5 seconds, it will continue with an invalid port. I am refactoring the port selection into the python script to solve this.